app/src/main/java/com/github/geohunt/app/data/repository/bounties/TeamsRepository.kt

```diff
@@ -0,0 +1,137 @@
+                    .setValue(true)
+                    .await()
+        }
+    }
+
+
+    override suspend fun leaveTeam(userId: String) {
+        val team = getUserTeam(userId).first() ?: return
+
+        withContext(ioDispatcher) {
+            teams.child(team.teamId).child("members").child(userId).removeValue()
+        }
+    }
+
+    override fun getTeam(teamId: String): Flow<Team> {
+        return teams.child(teamId).snapshots
+                .map { snapshotToTeam(it) }
+                .flowOn(ioDispatcher)
+    }
+
+    override suspend fun getUserTeam(): Flow<Team?> = getUserTeam(userRepository.getCurrentUser().id)
+
+    override fun getTeamScore(team: Team): Flow<Long> =
+        teams.child(team.teamId).child("score").snapshots
+            .map { it.getValue(Long::class.java) ?: throw TeamNotFoundException(team.teamId) }
+            .flowOn(ioDispatcher)
+
+    override fun getUserTeam(userId: String): Flow<Team?> {
+        return getTeams().map {
+            it.firstOrNull { team -> team.membersUid.contains(userId) }
+        }
+    }
+
+    override fun getTeams(): Flow<List<Team>> {
+        return teams.snapshots
+                .map { it -> it.children.map { snapshotToTeam(it) } }
+                .flowOn(ioDispatcher)
+    }
+
+    override suspend fun createTeam(name: String): Team {
+        return createTeam(name, userRepository.getCurrentUser().id)
+    }
+
+    override suspend fun createTeam(name: String, teamLeaderUid: String): Team {
+        return withContext(ioDispatcher) {
+            val newTeamReference = teams.push()
+            val teamId = newTeamReference.key!!
+
+            // set the leader
+            newTeamReference.child("name").setValue(name)
+            newTeamReference.child("score").setValue(0)
+            newTeamReference.child("teamLeader").setValue(teamLeaderUid)
+
+            joinTeam(teamId, teamLeaderUid)
+
+            Team(
+                teamId = teamId,
+                name = name,
+                membersUid = listOf(teamLeaderUid),
+                leaderUid = teamLeaderUid,
+                score = 0
+            )
+        }
+    }
+
+    override suspend fun deleteTeam(teamId: String) {
+        teams.child(teamId).removeValue().await()
+    }
+
+    /**
+     * Atomically update of the score
+     */
+    internal suspend fun atomicScoreAddAndAssign(team: Team, increment: Long) = withContext(ioDispatcher) {
+        teams.child(team.teamId)
+            .updateChildren(
+                hashMapOf("score" to ServerValue.increment(increment))
+            )
+            .await()
+    }
+
+    private fun snapshotToTeam(s: DataSnapshot): Team {
+        return Team(
+            teamId = s.key!!,
+            name = s.child("name").getValue(String::class.java)!!,
+            membersUid = s.child("members").toMap<Boolean>().filterValues { it }.keys.toList(),
+            leaderUid = s.child("teamLeader").getValue(String::class.java) ?: "",
```

Comment:
There has to be a team leader, maybe consider replacing ?: "" by !!

Reply:
That's a good point! ;)